evm/vm/evm.py

```diff
@@ -267,12 +267,58 @@ def mine_block(self, **mine_params):
         Mine the current block, applying
         """
         block = self.get_vm().mine_block(**mine_params)
+        self.validate_block(block)
         persist_block_to_db(self.db, block)
 
         self.header = self.create_header_from_parent(block.header)
 
         return block
 
+    def validate_block(self, block):
+        self.validate_seal(block.header)
+        self.validate_uncles(block)
+
+    def validate_uncles(self, block):
+        recent_ancestors = dict(
+            (ancestor.hash, ancestor)
+            for ancestor in self.get_ancestors(MAX_UNCLE_DEPTH+1))
+        recent_uncles = []
+        for ancestor in recent_ancestors.values():
+            recent_uncles.extend([uncle.hash for uncle in ancestor.uncles])
+        recent_ancestors[block.hash] = block
+        recent_uncles.append(block.hash)
+
+        for uncle in block.uncles:
+            if uncle.hash in recent_ancestors:
+                raise ValidationError(
+                    "Duplicate uncle: {0}".format(encode_hex(uncle.hash)))
+            recent_uncles.append(uncle.hash)
+
+            if uncle.hash in recent_ancestors:
+                raise ValidationError(
+                    "Uncle {0} cannot be an ancestor of {1}".format(
+                        encode_hex(uncle.hash), encode_hex(block.hash)))
+
+            if uncle.parent_hash not in recent_ancestors or (
+               uncle.parent_hash == block.header.parent_hash):
+                raise ValidationError(
+                    "Uncle's parent {0} is not an ancestor of {1}".format(
+                        encode_hex(uncle.parent_hash), encode_hex(block.hash)))
+
+            self.validate_seal(uncle)
+
+    def validate_seal(self, header):
+        check_pow(
+            header.block_number, header.mining_hash,
+            header.mix_hash, header.nonce, header.difficulty)
+
+    def get_ancestors(self, limit):
+        blocks = []
+        lower_limit = max(self.header.block_number - limit, 0)
+        for n in range(lower_limit, self.header.block_number):
+            blocks.insert(0, self.get_block_by_number(n))
```

Comment:
It is my understanding that this is one of the slowest operations you can do in python, inserting elements into position `0` of a list.  I'd suggest one of the following.

1. Reverse the range so that you can use `list.append(...)`
2. Use a `collections.deque` which has fast inserts to the head of lists.